### DISCUSSION TITLE
Difference between using jwt as access token and caching introspection response.

### DISCUSSION THREAD

**Discussion by errnesto** [zitadel/zitadel#6358] [Q&A]
I try to better understand the downsides of using jwt's instead of opaque access tokens.

The thing I always hear is that it is hard to revoke tokens (e.g. for logged out users)
But I really don't want to call the introspection endpoint on evey single request, so I have to cache the response.
So in the end let's say I cache the response for 5min what is the difference to a jwt with a lifetime of 5min?

I can also still introspect the jwt on more sesitive requests.

I can see that the token size is bigger, and depending on what information is included in the jwt, even an expired token can leak data…

Of course this is a tradeoff, but the downsides seem pretty small to me. I simply wan't to make sure I'm not missing something.

**Comment by fforootd** [comment]
Good question!

Short lived tokens diverge only a little from calling the introspect/userinfo endpoints. Main things to consider are the following:

- Introspect/Userinfo can update information (like access or user data) within the lifetime of a token so you always get the latest data available
- Token size can be an issue with chatty APIs or slow connections
- Token can expose information intended only for internal usage

Biggest issue with tokens though, is that upstream systems might still trust a token that is forwarded to them because of their valid signature. You certainly can work around this by verifying the token against a central service if your security needs are high enough.

My2cents to this are the following: 

If your security needs can handle a stale token or a potential mishap that a token is not verified to a central service then (short lived) tokens are totally fine. In a scalable architecture this gives you some benefits since you reduce the amount of calls to the central service by avoiding calls to introspect.

On the other hand, if you want to have a "by-default" more "secure" approach you should stick to opaque tokens and at some point build in caches where you explicitly want to have them.

So IMO it is better to explicitly "downgrade" security for availability/performance then the other way around

Hope this helps 😁
